### PR TITLE
fix(components): fixed add listeners not remove on beforeunmount

### PR DESCRIPTION
**Pull request by myronliu347**
涉及多个组件未在unmount 前移除事件监听， 导致内存泄漏
* popover
* tooltip
* focus-trap
* time-picker
* auto-complete

另外两处改动
1. vue 编译构建关闭 hoistStatic/cacheHandlers (目前这两处可能导致内存溢出)
2. v-click-outside 指令， startClick 用完后清空
3. 关闭官网文档 hoist-static 编译

### CODE REVIEW

File: packages/components/popper/src/trigger.vue [hunk 1]
```diff
@@ -135,6 +135,24 @@ onMounted(() => {
...
+      'onClick',
+      'onKeydown',
+      'onFocus',
+      'onBlur',
+      'onContextmenu',
+    ].forEach((eventName) => {
+      const handler = props[eventName]
+      if (handler) {
+        el.removeEventListener(eventName.slice(2).toLowerCase(), handler)
+      }
```
Comment (btea) [inline packages/components/popper/src/trigger.vue:152]: Event names in arrays may be written as `['mouseenter', 'mouseleave',...]`, and no conversion is required here.

Comment (myronliu347) [reply packages/components/popper/src/trigger.vue:144]: 这里是有必要，前面绑定了事件，最后都没移除掉

**Comment by myronliu347** [comment]
另外补充一些内存泄漏的相关信息， Vue teleport 外部组件切换的时候 （RouterView, Component is ）等都会造成内存泄漏的问题， 实际的例子可以参考官网的 popover组件 https://element-plus.org/zh-CN/component/popover.html
从这个页面切换出去之后，内存里还是会留着 dom, 
image

我在本地修复了事件回收，但是这个问题还在，移除掉 teleport 组件就正常，我试了下其他的组件库（Vuetify 等），使用了teleport 组件都存在这个问题， 建议不要使用官方的这个组件；

ps: vue issue里搜索到很多相关的内存的泄漏的问题，最后都没有一个准确的答复
https://github.com/vuejs/core/issues/11121

File: internal/build/src/tasks/modules.ts [hunk 2]
```diff
@@ -31,6 +31,12 @@ export const buildModules = async () => {
         plugins: {
           vue: vue({
             isProduction: true,
+            template: {
+              compilerOptions: {
+                hoistStatic: false,
+                cacheHandlers: false,
+              },
+            },
```
Comment (btea) [inline internal/build/src/tasks/modules.ts:39]: Is there any comparison data or scenario before and after the modification? In fact, I am not sure why these two configurations will cause memory leaks. 🤔

Comment (myronliu347) [reply internal/build/src/tasks/modules.ts:39]: https://github.com/vuejs/core/issues/5256
vue 官方就有， 三周前修复预计在3.5 才发布

Comment (myronliu347) [reply internal/build/src/tasks/modules.ts:39]: 数据化用 edge 浏览器，调试工具里有分离元素面板，你可以看下每次路由切换后有多少的内存泄漏

**Review by btea — APPROVED** [review]
I did a quick test and found that this modification is indeed effective.

File: internal/build/src/tasks/modules.ts (hunk 2 above)
Comment (namonakimono) [reply internal/build/src/tasks/modules.ts:39]: Hi btea, will you kindly review the changes and merge the PR if it meets the criteria of the repository?

The project here is quite heavy and the repeated memory leak causes severe problems than lightweight apps. If it is not merged, we need to fork the project and do some hack to use the fixed components.

Indeed, you could see how memory grows (up several million bytes) after click the table and virtualised table page several times (https://element-plus.org/en-US/component/table.html); by contrast, the ant design components have no these leak and the used memory can be recycled effectively.

**Comment by namonakimono** [comment]
@btea 

Hi btea, will you kindly review the changes and merge the PR if it meets the criteria of the repository?

The project we are working on is quite heavy and the repeated memory leak (esp. from popper) causes severe problems than lightweight apps. If it is not merged, we need to fork the project and do some hack to use the fixed components.

Indeed, you could see how memory grows after click the tree, table, and several other pages several times (https://element-plus.org/en-US/component/table.html).

Thank you very much.

image

image

image

File: packages/components/popper/src/trigger.vue [hunk 3]
```diff
@@ -58,6 +58,16 @@ const ariaExpanded = computed<string | undefined>(() => {
...
 
+const TRIGGER_ELE_EVENTS = [
+  'onMouseenter',
+  'onMouseleave',
+  'onClick',
+  'onKeydown',
+  'onFocus',
+  'onBlur',
+  'onContextmenu',
+]
```
Comment (tolking) [inline packages/components/popper/src/trigger.vue:69]:
```suggestion
] as const
```

File: internal/build/src/tasks/modules.ts (hunk 2 above)
Comment (tolking) [reply internal/build/src/tasks/modules.ts:39]: 或许可以增加一个 TODO，等以后 peerDependencies 的 vue: ^3.5.0 时恢复默认配置

**Comment by btea** [comment]
@tolking It looks like there has been no response for a long time, maybe we can help revise it.

**Comment by tolking** [comment]
@myronliu347 Do you have time to deal with relevant conflicts?

**Review by tolking — APPROVED** [review]

File: internal/build/src/tasks/modules.ts (hunk 2 above)
Comment (tolking) [reply internal/build/src/tasks/modules.ts:39]: 这里还需要修改 full-bundle.ts 文件

File: packages/components/autocomplete/src/autocomplete.vue [hunk 4]
```diff
@@ -350,10 +356,16 @@ const highlight = (index: number) => {
...
 
-onClickOutside(listboxRef, () => {
+const stopHandle = onClickOutside(listboxRef, () => {
   suggestionVisible.value && close()
 })
 
+onBeforeUnmount(() => {
+  if (stopHandle) {
+    stopHandle()
+  }
```
Comment (tolking) [inline packages/components/autocomplete/src/autocomplete.vue:366]:
```suggestion
  stopHandle?.()
```

File: packages/components/time-picker/src/common/picker.vue [hunk 5]
```diff
@@ -584,6 +585,12 @@ onClickOutside(actualInputRef, (e: PointerEvent) => {
   pickerVisible.value = false
 })
 
+onBeforeUnmount(() => {
+  if (stophandle) {
+    stophandle()
+  }
```
Comment (tolking) [inline packages/components/time-picker/src/common/picker.vue:591]:
```suggestion
  stophandle?.()
```

File: packages/components/tooltip/src/content.vue [hunk 6]
```diff
@@ -65,7 +63,8 @@ const { selector } = usePopperContainerId()
 const ns = useNamespace('tooltip')
 // TODO any is temporary, replace with `InstanceType<typeof ElPopperContent> | null` later
 const contentRef = ref<any>(null)
-const destroyed = ref(false)
+let stopHandle: ReturnType<typeof onClickOutside>
+// const destroyed = ref(false)
```
Comment (tolking) [inline packages/components/tooltip/src/content.vue:67]:
```suggestion
```

这里注释的代码可以不保留

File: packages/components/tooltip/src/content.vue [hunk 7]
```diff
@@ -91,7 +90,9 @@ const persistentRef = computed(() => {
 })
 
 onBeforeUnmount(() => {
-  destroyed.value = true
+  if (stopHandle) {
+    stopHandle()
+  }
```
Comment (tolking) [inline packages/components/tooltip/src/content.vue:95]:
```suggestion
  stopHandle?.()
```

**Review by tolking — COMMENTED** [review]
还需要修改 full-bundle.ts 文件中的 vue 配置

File: internal/build/src/tasks/full-bundle.ts [hunk 8]
```diff
@@ -39,6 +39,12 @@ async function buildFullEntry(minify: boolean) {
       plugins: {
         vue: vue({
           isProduction: true,
+          template: {
+            compilerOptions: {
+              hoistStatic: false,
+              cacheHandlers: false,
+            },
+          },
```
Comment (tolking) [inline internal/build/src/tasks/full-bundle.ts:47]: 相关的配置能不能增加一个 TODO，等以后 peerDependencies 的 vue: ^3.5.0 后移除

Comment (myronliu347) [reply internal/build/src/tasks/full-bundle.ts:47]: 能升vue的时候 vite 都升级了吧

Comment (myronliu347) [reply internal/build/src/tasks/full-bundle.ts:47]: 这个关闭对基础组件开发没有的影响的，开启可能因为vue 本身编译的bug 导致一些问题，副作用更大

**Review by tolking — APPROVED** [review]

**Comment by myronliu347** [comment]
麻烦尽快合一下，

**Pull request merged**